// @teambit/legacy-bit-id@1.0.1/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidScopeNameFromRemote = exports.InvalidScopeName = exports.InvalidName = exports.isValidIdChunk = exports.isValidScopeName = exports.VERSION_DELIMITER = exports.BitId = void 0;
const bit_id_1 = __importStar(require("./bit-id"));
exports.BitId = bit_id_1.default;
Object.defineProperty(exports, "VERSION_DELIMITER", { enumerable: true, get: function () { return bit_id_1.VERSION_DELIMITER; } });
const exceptions_1 = require("./exceptions");
Object.defineProperty(exports, "InvalidName", { enumerable: true, get: function () { return exceptions_1.InvalidName; } });
Object.defineProperty(exports, "InvalidScopeName", { enumerable: true, get: function () { return exceptions_1.InvalidScopeName; } });
Object.defineProperty(exports, "InvalidScopeNameFromRemote", { enumerable: true, get: function () { return exceptions_1.InvalidScopeNameFromRemote; } });
const is_valid_scope_name_1 = __importDefault(require("./utils/is-valid-scope-name"));
exports.isValidScopeName = is_valid_scope_name_1.default;
const is_valid_id_chunk_1 = __importDefault(require("./utils/is-valid-id-chunk"));
exports.isValidIdChunk = is_valid_id_chunk_1.default;
