// @bitdev/react.generators.react-templates@1.1.12/dist/webpack-template/app-template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactWebpackTemplate = void 0;
const index_file_1 = require("./files/index-file");
const app_1 = require("./files/app");
const react_app_1 = require("./files/react-app");
const composition_1 = require("./files/composition");
const app_root_1 = require("./files/app-root");
const server_root_1 = require("./files/server-root");
const index_html_1 = require("./files/index-html");
const docs_1 = require("./files/docs");
class ReactWebpackTemplate {
    constructor(name = 'react-cjs-app', description = 'create a webpack-based React app with SSR for CJS support', hidden = false, env, installMissingDependencies = true, isApp = true) {
        this.name = name;
        this.description = description;
        this.hidden = hidden;
        this.env = env;
        this.installMissingDependencies = installMissingDependencies;
        this.isApp = isApp;
    }
    generateFiles(context) {
        return [
            (0, index_file_1.indexFile)(context),
            (0, composition_1.compositionFile)(context),
            (0, app_1.AppFile)(context),
            (0, react_app_1.reactAppFile)(context),
            (0, app_root_1.AppRootFile)(context),
            (0, server_root_1.serverRootFile)(context),
            (0, docs_1.appDocsFile)(context),
            (0, index_html_1.indexHtml)(),
        ];
    }
    static from(options = {}) {
        return () => new ReactWebpackTemplate(options.name, options.description, options.hidden, options.env);
    }
}
exports.ReactWebpackTemplate = ReactWebpackTemplate;
