// @bitdev/react.generators.react-templates@1.1.12/dist/context-template/context-template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactContextTemplate = void 0;
const index_file_1 = require("./files/index-file");
const context_provider_1 = require("./files/context-provider");
const context_1 = require("./files/context");
const docs_1 = require("./files/docs");
const test_1 = require("./files/test");
const use_context_1 = require("./files/use-context");
const composition_1 = require("./files/composition");
class ReactContextTemplate {
    constructor(name = 'react-context', description = 'A react context component. ', hidden = false, env) {
        this.name = name;
        this.description = description;
        this.hidden = hidden;
        this.env = env;
    }
    generateFiles(context) {
        return [
            (0, index_file_1.indexFile)(context),
            (0, docs_1.docsFile)(context),
            (0, composition_1.compositionFile)(context),
            (0, context_1.contextFile)(context),
            (0, test_1.testFile)(context),
            (0, use_context_1.useContextFile)(context),
            (0, context_provider_1.contextProviderFile)(context),
        ];
    }
    static from(options = {}) {
        return () => new ReactContextTemplate(options.name, options.description, options.hidden, options.env);
    }
}
exports.ReactContextTemplate = ReactContextTemplate;
