// @teambit/react.generator.react-starters@1.0.6/starters/react-workspace/react-workspace-starter.ts
import { WorkspaceContext, WorkspaceTemplate } from '@teambit/generator';
import { generateFiles as generateCommonFiles } from '../../workspace-common';
import { FORKED_ENV_NAME } from '../../workspace-common/constants';

export type ReactComponentTemplateOptions = {
  /**
   * name of the template
   */
  name?: string;

  /**
   * description of the template.
   */
  description?: string;

  /**
   * hide the template from the `bit templates` command.
   */
  hidden?: boolean;
};

export class ReactWorkspaceStarter implements WorkspaceTemplate {
  constructor(
    readonly name = 'react',
    readonly description = 'React workspace with demo components',
    readonly hidden = false
  ) {}

  async generateFiles(context: WorkspaceContext) {
    return generateCommonFiles(context);
  }

  fork() {
    return [
      {
        id: 'teambit.react/react-env-extension',
        targetName: FORKED_ENV_NAME,
      },
    ];
  }

  static from(options: ReactComponentTemplateOptions = {}) {
    return () =>
      new ReactWorkspaceStarter(
        options.name,
        options.description,
        options.hidden
      );
  }
}
