// @teambit/webpack.webpack-bundler@1.0.7/webpack-bundler.ts
import { BitError } from '@teambit/bit-error';
import type {
  Bundler,
  BundlerResult,
  Asset,
  Target,
  EntriesAssetsMap,
  BundlerContextMetaData,
} from '@teambit/bundler';
import type { Logger } from '@teambit/logger';
import { compact, isEmpty } from 'lodash';
import mapSeries from 'p-map-series';
import type {
  webpack as webpackCompiler,
  Compiler,
  Configuration,
  StatsCompilation,
  StatsAsset,
} from 'webpack';
import { sep } from 'path';
import { EnvContext } from '@teambit/envs';
import {
  GlobalWebpackConfigTransformContext,
  WebpackConfigMutator,
} from '@teambit/webpack';
import { PubsubMain } from '@teambit/pubsub';
import { WebpackBundlerOptions } from './webpack-bundler-options';
import { generateTransformers, runTransformers } from './transformers';
import { configFactory } from './config/webpack.config';

export type BundlerAppContext = {
  logger: Logger;
};

type WebpackBundlerAspectsContext = {
  logger: Logger;
  workspacePath?: string;
  pubsub?: PubsubMain;
};

type AssetsMap = { [assetId: string]: Asset };
export class WebpackBundler implements Bundler {
  private readonly webpack: typeof webpackCompiler;

  constructor(
    /**
     * targets to bundle.
     */
    private targets: Target[],

    /**
     * webpack configuration.
     */
    private configs: Configuration[],

    private logger: Logger,

    private webpackModulePath: string = require.resolve('webpack'),

    private metaData?: BundlerContextMetaData | undefined
  ) {
    // eslint-disable-next-line import/no-dynamic-require, global-require
    this.webpack = require(this.webpackModulePath);
  }

  async run(): Promise<BundlerResult[]> {
    const startTime = Date.now();
    const compilers = this.configs.map((config: any) => this.webpack(config));
    const initiator = this.metaData?.initiator;
    const envId = this.metaData?.envId;
    const initiatorMessage = initiator
      ? `process initiated by: ${initiator}.`
      : '';
    const envIdMessage = envId ? `config created by env: ${envId}.` : '';

    const longProcessLogger = this.logger.createLongProcessLogger(
      'running Webpack bundler',
      compilers.length
    );
    const componentOutput = await mapSeries(compilers, (compiler: Compiler) => {
      const components = this.getComponents(compiler.outputPath);
      const componentsLengthMessage = `running on ${components.length} components`;
      const fullMessage = `${initiatorMessage} ${envIdMessage} ${componentsLengthMessage}`;
      const ids = components
        .map((component) => component.id.toString())
        .join(', ');
      longProcessLogger.logProgress(`${fullMessage}`);
      this.logger.debug(`${fullMessage}\ncomponents ids: ${ids}`);
      return new Promise((resolve) =>
        // TODO: split to multiple processes to reduce time and configure concurrent builds.
        // @see https://github.com/trivago/parallel-webpack
        compiler.run((err, stats) => {
          if (err) {
            this.logger.error(
              'get error from webpack compiler, full error:',
              err
            );

            return resolve({
              errors: [`${err.toString()}\n${err.stack}`],
              components,
            });
          }
          if (!stats) throw new BitError('unknown build error');
          // const info = stats.toJson();

          const info = stats.toJson({
            all: false,
            entrypoints: true,
            warnings: true,
            errors: true,
            assets: true,
            chunkGroupAuxiliary: true,
            relatedAssets: true,
            cachedAssets: true,
          });
          const assetsMap = this.getAssets(info);
          const entriesAssetsMap = this.getEntriesAssetsMap(info, assetsMap);

          return resolve({
            assets: Object.values(assetsMap),
            assetsByChunkName: info.assetsByChunkName,
            entriesAssetsMap,
            errors: this.getErrors(info),
            outputPath: stats.compilation.outputOptions.path,
            components,
            warnings: info.warnings,
            startTime,
            endTime: Date.now(),
          });
        })
      );
    });
    longProcessLogger.end();
    return componentOutput as BundlerResult[];
  }

  private getErrors(stats: StatsCompilation): Error[] {
    if (!stats.errors) return [];
    const fieldsToShow = [
      'message',
      'moduleId',
      'moduleName',
      'moduleIdentifier',
      'loc',
    ];
    return stats.errors.map((webpackError) => {
      const lines = fieldsToShow.map((fieldName) => {
        if (webpackError[fieldName]) {
          return `${fieldName}: ${webpackError[fieldName]}`;
        }
        return undefined;
      });
      const errorMessage = compact(lines).join('\n');
      return new BitError(errorMessage);
    });
  }

  private getAssets(stats: StatsCompilation): AssetsMap {
    if (!stats.assets) return {};
    return stats.assets.reduce((acc, asset) => {
      acc[asset.name] = {
        name: asset.name,
        size: asset.size,
        compressedSize: this.getCompressedSize(asset),
      };
      return acc;
    }, {});
  }

  private getEntriesAssetsMap(
    stats: StatsCompilation,
    assetsMap: AssetsMap
  ): EntriesAssetsMap {
    const entriesMap = stats.entrypoints;
    if (!entriesMap || !Object.keys(assetsMap).length) return {};
    Object.entries(entriesMap).forEach(([, entryVal]) => {
      let compressedAssetsSize = 0;
      let compressedAuxiliaryAssetsSize = 0;
      entryVal.assets?.forEach((asset) => {
        const compressedSize = assetsMap[asset.name]?.compressedSize;
        if (compressedSize) {
          // @ts-ignore
          asset.compressedSize = compressedSize; // eslint-disable-line no-param-reassign
          compressedAssetsSize += compressedSize;
        }
      });
      entryVal.auxiliaryAssets?.forEach((asset) => {
        const compressedSize = assetsMap[asset.name]?.compressedSize;
        if (compressedSize) {
          // @ts-ignore
          asset.compressedSize = compressedSize; // eslint-disable-line no-param-reassign
          compressedAuxiliaryAssetsSize += compressedSize;
        }
      });
      // @ts-ignore
      entryVal.compressedAssetsSize = compressedAssetsSize; // eslint-disable-line no-param-reassign
      // @ts-ignore
      entryVal.compressedAuxiliaryAssetsSize = compressedAuxiliaryAssetsSize; // eslint-disable-line no-param-reassign
    });
    return entriesMap as any as EntriesAssetsMap;
  }

  private getCompressedSize(asset: StatsAsset): number | undefined {
    if (!asset.related || isEmpty(asset.related)) return undefined;
    const gzipped = asset.related.find(
      (relatedAsset) => relatedAsset.type === 'gzipped'
    );
    if (!gzipped) return undefined;
    return gzipped.size;
  }

  private getComponents(outputPath: string) {
    const path = outputPath.substring(0, outputPath.lastIndexOf(sep));
    const target = this.targets.find(
      (targetCandidate) => path === targetCandidate.outputPath
    );

    if (!target)
      throw new Error(`Could not find component id for path "${path}"`);
    return target.components;
  }

  static from(options: WebpackBundlerOptions) {
    return (context: EnvContext) => {
      const logger = context.createLogger(options.name || 'webpack-bundler');
      const bundlerContext: WebpackBundlerAspectsContext = {
        logger,
      };

      return WebpackBundler.create(options, bundlerContext);
    };
  }

  static create(
    options: WebpackBundlerOptions,
    aspectContext: WebpackBundlerAspectsContext
  ) {
    const { logger } = aspectContext;
    const { transformers } = options;
    const transformerContext: GlobalWebpackConfigTransformContext = {
      mode: 'prod',
      isEnvTemplate: options.bundlerContext.metaData?.isEnvTemplate
    };
    const configs =
      options.targets.map((target) => {
        const finalTransformerContext = { ...transformerContext, target };
        const baseConfig = configFactory(target, options.bundlerContext);
        const internalTransformers = generateTransformers(
          logger,
          finalTransformerContext,
          undefined,
          target
        );
        const configMutator = new WebpackConfigMutator(baseConfig);

        const afterMutation = runTransformers(
          configMutator.clone(),
          [...internalTransformers, ...transformers],
          finalTransformerContext
        );
        return afterMutation.raw;
      });

    return new WebpackBundler(
      options.targets,
      configs,
      logger,
      options.webpackModulePath
    );
  }
}
