// @teambit/react.webpack.react-webpack@1.0.23/dist/webpack.config.component.prod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.configComponentProdFactory = void 0;
const component_id_1 = require("@teambit/component-id");
// Make sure the bit-react-transformer is a dependency
// TODO: remove it once we can set policy from component to component then set it via the component.json
require("@teambit/react.babel.bit-react-transformer");
// This is the production and development configuration.
// It is focused on developer experience, fast rebuilds, and a minimal bundle.
// eslint-disable-next-line complexity
function configComponentProdFactory() {
    return {
        module: {
            rules: [
                {
                    test: /\.js$/,
                    include: [/node_modules/, /\/dist\//],
                    exclude: /@teambit\/legacy/,
                    descriptionData: { componentId: component_id_1.ComponentID.isValidObject },
                    use: [
                        {
                            loader: require.resolve('babel-loader'),
                            options: {
                                sourceType: 'unambiguous',
                                babelrc: false,
                                configFile: false,
                                plugins: [
                                    // for component highlighting in preview.
                                    [
                                        require.resolve('@teambit/react.babel.bit-react-transformer'),
                                    ],
                                ],
                                // turn off all optimizations (only slow down for node_modules)
                                compact: false,
                                minified: false,
                            },
                        },
                    ],
                },
            ],
        },
    };
}
exports.configComponentProdFactory = configComponentProdFactory;
