// @teambit/webpack.webpack-bundler@1.0.7/dist/webpack-bundler.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebpackBundler = void 0;
const bit_error_1 = require("@teambit/bit-error");
const lodash_1 = require("lodash");
const p_map_series_1 = __importDefault(require("p-map-series"));
const path_1 = require("path");
const webpack_1 = require("@teambit/webpack");
const transformers_1 = require("./transformers");
const webpack_config_1 = require("./config/webpack.config");
class WebpackBundler {
    constructor(
    /**
     * targets to bundle.
     */
    targets, 
    /**
     * webpack configuration.
     */
    configs, logger, webpackModulePath = require.resolve('webpack'), metaData) {
        this.targets = targets;
        this.configs = configs;
        this.logger = logger;
        this.webpackModulePath = webpackModulePath;
        this.metaData = metaData;
        // eslint-disable-next-line import/no-dynamic-require, global-require
        this.webpack = require(this.webpackModulePath);
    }
    run() {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const startTime = Date.now();
            const compilers = this.configs.map((config) => this.webpack(config));
            const initiator = (_a = this.metaData) === null || _a === void 0 ? void 0 : _a.initiator;
            const envId = (_b = this.metaData) === null || _b === void 0 ? void 0 : _b.envId;
            const initiatorMessage = initiator
                ? `process initiated by: ${initiator}.`
                : '';
            const envIdMessage = envId ? `config created by env: ${envId}.` : '';
            const longProcessLogger = this.logger.createLongProcessLogger('running Webpack bundler', compilers.length);
            const componentOutput = yield (0, p_map_series_1.default)(compilers, (compiler) => {
                const components = this.getComponents(compiler.outputPath);
                const componentsLengthMessage = `running on ${components.length} components`;
                const fullMessage = `${initiatorMessage} ${envIdMessage} ${componentsLengthMessage}`;
                const ids = components
                    .map((component) => component.id.toString())
                    .join(', ');
                longProcessLogger.logProgress(`${fullMessage}`);
                this.logger.debug(`${fullMessage}\ncomponents ids: ${ids}`);
                return new Promise((resolve) => 
                // TODO: split to multiple processes to reduce time and configure concurrent builds.
                // @see https://github.com/trivago/parallel-webpack
                compiler.run((err, stats) => {
                    if (err) {
                        this.logger.error('get error from webpack compiler, full error:', err);
                        return resolve({
                            errors: [`${err.toString()}\n${err.stack}`],
                            components,
                        });
                    }
                    if (!stats)
                        throw new bit_error_1.BitError('unknown build error');
                    // const info = stats.toJson();
                    const info = stats.toJson({
                        all: false,
                        entrypoints: true,
                        warnings: true,
                        errors: true,
                        assets: true,
                        chunkGroupAuxiliary: true,
                        relatedAssets: true,
                        cachedAssets: true,
                    });
                    const assetsMap = this.getAssets(info);
                    const entriesAssetsMap = this.getEntriesAssetsMap(info, assetsMap);
                    return resolve({
                        assets: Object.values(assetsMap),
                        assetsByChunkName: info.assetsByChunkName,
                        entriesAssetsMap,
                        errors: this.getErrors(info),
                        outputPath: stats.compilation.outputOptions.path,
                        components,
                        warnings: info.warnings,
                        startTime,
                        endTime: Date.now(),
                    });
                }));
            });
            longProcessLogger.end();
            return componentOutput;
        });
    }
    getErrors(stats) {
        if (!stats.errors)
            return [];
        const fieldsToShow = [
            'message',
            'moduleId',
            'moduleName',
            'moduleIdentifier',
            'loc',
        ];
        return stats.errors.map((webpackError) => {
            const lines = fieldsToShow.map((fieldName) => {
                if (webpackError[fieldName]) {
                    return `${fieldName}: ${webpackError[fieldName]}`;
                }
                return undefined;
            });
            const errorMessage = (0, lodash_1.compact)(lines).join('\n');
            return new bit_error_1.BitError(errorMessage);
        });
    }
    getAssets(stats) {
        if (!stats.assets)
            return {};
        return stats.assets.reduce((acc, asset) => {
            acc[asset.name] = {
                name: asset.name,
                size: asset.size,
                compressedSize: this.getCompressedSize(asset),
            };
            return acc;
        }, {});
    }
    getEntriesAssetsMap(stats, assetsMap) {
        const entriesMap = stats.entrypoints;
        if (!entriesMap || !Object.keys(assetsMap).length)
            return {};
        Object.entries(entriesMap).forEach(([, entryVal]) => {
            var _a, _b;
            let compressedAssetsSize = 0;
            let compressedAuxiliaryAssetsSize = 0;
            (_a = entryVal.assets) === null || _a === void 0 ? void 0 : _a.forEach((asset) => {
                var _a;
                const compressedSize = (_a = assetsMap[asset.name]) === null || _a === void 0 ? void 0 : _a.compressedSize;
                if (compressedSize) {
                    // @ts-ignore
                    asset.compressedSize = compressedSize; // eslint-disable-line no-param-reassign
                    compressedAssetsSize += compressedSize;
                }
            });
            (_b = entryVal.auxiliaryAssets) === null || _b === void 0 ? void 0 : _b.forEach((asset) => {
                var _a;
                const compressedSize = (_a = assetsMap[asset.name]) === null || _a === void 0 ? void 0 : _a.compressedSize;
                if (compressedSize) {
                    // @ts-ignore
                    asset.compressedSize = compressedSize; // eslint-disable-line no-param-reassign
                    compressedAuxiliaryAssetsSize += compressedSize;
                }
            });
            // @ts-ignore
            entryVal.compressedAssetsSize = compressedAssetsSize; // eslint-disable-line no-param-reassign
            // @ts-ignore
            entryVal.compressedAuxiliaryAssetsSize = compressedAuxiliaryAssetsSize; // eslint-disable-line no-param-reassign
        });
        return entriesMap;
    }
    getCompressedSize(asset) {
        if (!asset.related || (0, lodash_1.isEmpty)(asset.related))
            return undefined;
        const gzipped = asset.related.find((relatedAsset) => relatedAsset.type === 'gzipped');
        if (!gzipped)
            return undefined;
        return gzipped.size;
    }
    getComponents(outputPath) {
        const path = outputPath.substring(0, outputPath.lastIndexOf(path_1.sep));
        const target = this.targets.find((targetCandidate) => path === targetCandidate.outputPath);
        if (!target)
            throw new Error(`Could not find component id for path "${path}"`);
        return target.components;
    }
    static from(options) {
        return (context) => {
            const logger = context.createLogger(options.name || 'webpack-bundler');
            const bundlerContext = {
                logger,
            };
            return WebpackBundler.create(options, bundlerContext);
        };
    }
    static create(options, aspectContext) {
        var _a;
        const { logger } = aspectContext;
        const { transformers } = options;
        const transformerContext = {
            mode: 'prod',
            isEnvTemplate: (_a = options.bundlerContext.metaData) === null || _a === void 0 ? void 0 : _a.isEnvTemplate
        };
        const configs = options.targets.map((target) => {
            const finalTransformerContext = Object.assign(Object.assign({}, transformerContext), { target });
            const baseConfig = (0, webpack_config_1.configFactory)(target, options.bundlerContext);
            const internalTransformers = (0, transformers_1.generateTransformers)(logger, finalTransformerContext, undefined, target);
            const configMutator = new webpack_1.WebpackConfigMutator(baseConfig);
            const afterMutation = (0, transformers_1.runTransformers)(configMutator.clone(), [...internalTransformers, ...transformers], finalTransformerContext);
            return afterMutation.raw;
        });
        return new WebpackBundler(options.targets, configs, logger, options.webpackModulePath);
    }
}
exports.WebpackBundler = WebpackBundler;
