// @bitdev/react.generators.react-templates@1.1.12/dist/theme-template/theme-template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThemeTemplate = void 0;
const index_file_1 = require("./files/index-file");
const composition_file_1 = require("./files/composition-file");
const provider_file_1 = require("./files/provider-file");
const docs_file_1 = require("./files/docs-file");
const theme_styles_1 = require("./files/theme-styles");
const tokens_file_1 = require("./files/tokens-file");
const theme_spec_1 = require("./files/theme-spec");
const theme_file_1 = require("./files/theme-file");
class ThemeTemplate {
    constructor(name = 'theme', description = 'create a React theme supporting CSS variables and CSS in JS', hidden = false, env, installMissingDependencies = true) {
        this.name = name;
        this.description = description;
        this.hidden = hidden;
        this.env = env;
        this.installMissingDependencies = installMissingDependencies;
    }
    generateFiles(context) {
        return [
            (0, index_file_1.themeIndexFile)(context),
            (0, composition_file_1.themeCompositionFile)(context),
            (0, provider_file_1.providerFile)(context),
            (0, theme_file_1.themeFile)(context),
            (0, docs_file_1.themeDocsFile)(context),
            (0, theme_styles_1.themeStyles)(context),
            (0, tokens_file_1.themeTokenFile)(context),
            (0, theme_spec_1.specFile)(context),
        ];
    }
    static from(options = {}) {
        return () => new ThemeTemplate(options.name, options.description, options.hidden, options.env);
    }
}
exports.ThemeTemplate = ThemeTemplate;
