// @teambit/webpack.webpack-bundler@1.0.7/dist/config/webpack.config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.configFactory = void 0;
/* eslint-disable complexity */
const webpack_1 = __importDefault(require("webpack"));
const lodash_1 = require("lodash");
const compression_webpack_plugin_1 = __importDefault(require("compression-webpack-plugin"));
const path_1 = require("path");
const html_webpack_plugin_1 = __importDefault(require("html-webpack-plugin"));
const webpack_fallbacks_provide_plugin_config_1 = require("./webpack-fallbacks-provide-plugin-config");
const webpack_fallbacks_aliases_1 = require("./webpack-fallbacks-aliases");
const webpack_fallbacks_1 = require("./webpack-fallbacks");
function generateHtmlPlugin(config) {
    const baseConfig = {
        filename: config.filename,
        chunks: config.chunks,
        chunksSortMode: config.chunkOrder,
        title: config.title,
        templateContent: config.templateContent,
        minify: config.minify,
        cache: false,
        favicon: config.favicon,
    };
    const filteredConfig = (0, lodash_1.omitBy)(baseConfig, lodash_1.isUndefined);
    return new html_webpack_plugin_1.default(filteredConfig);
}
function generateHtmlPlugins(configs) {
    return configs.map((config) => generateHtmlPlugin(config));
}
function configFactory(target, context) {
    var _a, _b, _c;
    let truthyEntries = Array.isArray(target.entries) && target.entries.length
        ? target.entries.filter(Boolean)
        : target.entries || {};
    if (Array.isArray(truthyEntries) && !truthyEntries.length) {
        truthyEntries = {};
    }
    const dev = Boolean(context.development);
    const htmlConfig = (_a = target.html) !== null && _a !== void 0 ? _a : context.html;
    const compress = (_b = target.compress) !== null && _b !== void 0 ? _b : context.compress;
    const htmlPlugins = htmlConfig ? generateHtmlPlugins(htmlConfig) : undefined;
    const splitChunks = (_c = target.chunking) === null || _c === void 0 ? void 0 : _c.splitChunks;
    const config = {
        mode: dev ? 'development' : 'production',
        // Stop compilation early in production
        bail: true,
        // These are the "entry points" to our application.
        // This means they will be the "root" imports that are included in JS bundle.
        // @ts-ignore
        entry: truthyEntries,
        infrastructureLogging: {
            level: 'error',
        },
        output: {
            // The build folder.
            path: `${target.outputPath}${path_1.sep}public`,
        },
        stats: {
            errorDetails: true,
        },
        resolve: {
            // TODO - check - we should not need both fallbacks and alias and provider plugin
            alias: webpack_fallbacks_aliases_1.fallbacksAliases,
            fallback: webpack_fallbacks_1.fallbacks,
        },
        plugins: [
            new webpack_1.default.ProvidePlugin(webpack_fallbacks_provide_plugin_config_1.fallbacksProvidePluginConfig)
        ],
    };
    if (target.filename) {
        config.output = config.output || {};
        config.output.filename = target.filename;
    }
    if (target.chunkFilename) {
        config.output = config.output || {};
        config.output.chunkFilename = target.chunkFilename;
    }
    if (target.runtimeChunkName) {
        config.optimization = config.optimization || {};
        config.optimization.runtimeChunk = {
            name: target.runtimeChunkName,
        };
    }
    if (splitChunks) {
        config.optimization = config.optimization || {};
        config.optimization.splitChunks = {
            chunks: 'all',
            name: false,
        };
    }
    if (htmlPlugins && htmlPlugins.length) {
        if (!config.plugins) {
            config.plugins = [];
        }
        config.plugins = config.plugins.concat(htmlPlugins);
    }
    if (compress) {
        if (!config.plugins) {
            config.plugins = [];
        }
        config.plugins = config.plugins.concat(new compression_webpack_plugin_1.default());
    }
    return config;
}
exports.configFactory = configFactory;
