// @teambit/webpack.webpack-dev-server@1.0.7/dist/webpack-dev-server.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebpackDevServer = void 0;
const path_1 = require("path");
const find_root_1 = __importDefault(require("find-root"));
const util_1 = require("util");
const webpack_1 = require("@teambit/webpack");
const webpack_webpack_bundler_1 = require("@teambit/webpack.webpack-bundler");
const object_hash_1 = __importDefault(require("object-hash"));
const webpack_dev_config_1 = require("./webpack.dev.config");
class WebpackDevServer {
    constructor(config, webpackModulePath = require.resolve('webpack'), webpackDevServerModulePath = require.resolve('webpack-dev-server')) {
        this.config = config;
        this.webpackModulePath = webpackModulePath;
        this.webpackDevServerModulePath = webpackDevServerModulePath;
        this.id = 'webpack-dev-server';
        // eslint-disable-next-line react/static-property-placement
        this.displayName = 'Webpack dev server';
        // eslint-disable-next-line import/no-dynamic-require, global-require
        this.webpack = require(this.webpackModulePath);
        // eslint-disable-next-line import/no-dynamic-require, global-require
        this.WsDevServer = require(this.webpackDevServerModulePath);
    }
    getCompiler() {
        return this.webpack(this.config);
    }
    version() {
        // Resolve version from the webpack-dev-server package.json
        try {
            const root = (0, find_root_1.default)(this.webpackDevServerModulePath);
            const packageJsonPath = (0, path_1.join)(root, 'package.json');
            // eslint-disable-next-line import/no-dynamic-require, global-require
            const packageJson = require(packageJsonPath);
            return packageJson.version;
        }
        catch (err) {
            return 'unknown';
        }
    }
    displayConfig() {
        return (0, util_1.inspect)(this.config, { depth: 10 });
    }
    hash() {
        const objToHash = {
            webpack: this.webpackModulePath,
            webpackDevServer: this.webpackDevServerModulePath,
            config: this.config,
        };
        return (0, object_hash_1.default)(objToHash);
    }
    listen(port) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.config.devServer) {
                throw new Error('Missing devServer configuration for webpack');
            }
            // Prevent different port between the config port and the listen arg port
            this.config.devServer.port = port;
            // Adding signal listeners to make sure we immediately close the process on sigint / sigterm (otherwise webpack dev server closing will take time)
            this.addSignalListener();
            // Compatibility check for Webpack dev server v3 (e.g. for Angular v8)
            if (typeof this.WsDevServer.addDevServerEntrypoints !== 'undefined') {
                // @ts-ignore in the capsules it throws an error about compatibilities issues between webpack.compiler and webpackDevServer/webpack/compiler
                const webpackDs = new this.WsDevServer(this.getCompiler(), this.config.devServer);
                return webpackDs.listen(port);
            }
            // @ts-ignore in the capsules it throws an error about compatibilities issues between webpack.compiler and webpackDevServer/webpack/compiler
            const webpackDs = new this.WsDevServer(this.config.devServer, this.getCompiler());
            yield webpackDs.start();
            return webpackDs.server;
        });
    }
    addSignalListener() {
        process.on('SIGTERM', () => {
            process.exit();
        });
        process.on('SIGINT', () => {
            process.exit();
        });
    }
    static from(options) {
        return (context) => {
            var _a;
            const name = options.name || 'webpack-dev-server';
            const logger = context.createLogger(name);
            const workspacePath = (_a = context.getAspect('teambit.workspace/workspace')) === null || _a === void 0 ? void 0 : _a.path;
            const pubsub = context.getAspect('teambit.harmony/pubsub');
            const aspectContext = {
                logger,
                workspacePath,
                pubsub,
            };
            return WebpackDevServer.create(options, aspectContext);
        };
    }
    static create(options, aspectContext) {
        const { logger, workspacePath, pubsub } = aspectContext;
        const { devServerContext } = options;
        const config = (0, webpack_dev_config_1.configFactory)(options.devServerContext.id, workspacePath, devServerContext.entry, devServerContext.rootPath, devServerContext.publicPath, pubsub);
        const configMutator = new webpack_1.WebpackConfigMutator(config);
        const transformerContext = Object.assign(devServerContext, { mode: 'dev' });
        const internalTransformers = (0, webpack_webpack_bundler_1.generateTransformers)(logger, undefined, transformerContext);
        const afterMutation = (0, webpack_webpack_bundler_1.runTransformers)(configMutator.clone(), [...internalTransformers, ...(options.transformers || [])], transformerContext);
        return new WebpackDevServer(afterMutation.raw, options.webpackModulePath, options.webpackDevServerModulePath);
    }
}
exports.WebpackDevServer = WebpackDevServer;
