// @teambit/documenter.markdown.heading@0.1.8/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.h6 = exports.h5 = exports.h4 = exports.h3 = exports.h2 = exports.h1 = exports.createHeading = void 0;
var create_heading_1 = require("./create-heading");
Object.defineProperty(exports, "createHeading", { enumerable: true, get: function () { return create_heading_1.createHeading; } });
Object.defineProperty(exports, "h1", { enumerable: true, get: function () { return create_heading_1.h1; } });
Object.defineProperty(exports, "h2", { enumerable: true, get: function () { return create_heading_1.h2; } });
Object.defineProperty(exports, "h3", { enumerable: true, get: function () { return create_heading_1.h3; } });
Object.defineProperty(exports, "h4", { enumerable: true, get: function () { return create_heading_1.h4; } });
Object.defineProperty(exports, "h5", { enumerable: true, get: function () { return create_heading_1.h5; } });
Object.defineProperty(exports, "h6", { enumerable: true, get: function () { return create_heading_1.h6; } });
