// @teambit/react.generator.react-starters@1.0.6/dist/starters/react-workspace/react-workspace-starter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactWorkspaceStarter = void 0;
const workspace_common_1 = require("../../workspace-common");
const constants_1 = require("../../workspace-common/constants");
class ReactWorkspaceStarter {
    constructor(name = 'react', description = 'React workspace with demo components', hidden = false) {
        this.name = name;
        this.description = description;
        this.hidden = hidden;
    }
    generateFiles(context) {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, workspace_common_1.generateFiles)(context);
        });
    }
    fork() {
        return [
            {
                id: 'teambit.react/react-env-extension',
                targetName: constants_1.FORKED_ENV_NAME,
            },
        ];
    }
    static from(options = {}) {
        return () => new ReactWorkspaceStarter(options.name, options.description, options.hidden);
    }
}
exports.ReactWorkspaceStarter = ReactWorkspaceStarter;
