// @teambit/preview.react-preview@1.0.23/dist/react-preview.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactPreview = void 0;
const object_hash_1 = __importDefault(require("object-hash"));
// Make sure bit recognizes the dependencies
require("webpack-dev-server");
require("webpack");
const webpack_webpack_bundler_1 = require("@teambit/webpack.webpack-bundler");
const webpack_webpack_dev_server_1 = require("@teambit/webpack.webpack-dev-server");
const dev_server_transformer_1 = require("./dev-server-transformer");
const bundler_transformers_1 = require("./bundler-transformers");
const template_transformers_1 = require("./template-transformers");
class ReactPreview {
    constructor(name, mounter = require.resolve('./mounter'), docsTemplate = require.resolve('./docs'), previewConfig = {}, transformers = [], workspace, hostDependencies, devServer, bundler, webpackModulePath = require.resolve('webpack'), webpackDevServerModulePath = require.resolve('webpack-dev-server')) {
        this.name = name;
        this.mounter = mounter;
        this.docsTemplate = docsTemplate;
        this.previewConfig = previewConfig;
        this.transformers = transformers;
        this.workspace = workspace;
        this.hostDependencies = hostDependencies;
        this.devServer = devServer;
        this.bundler = bundler;
        this.webpackModulePath = webpackModulePath;
        this.webpackDevServerModulePath = webpackDevServerModulePath;
    }
    getDevServer(context) {
        if (this.devServer)
            return () => this.devServer;
        return webpack_webpack_dev_server_1.WebpackDevServer.from({
            webpackModulePath: this.webpackModulePath,
            webpackDevServerModulePath: this.webpackDevServerModulePath,
            devServerContext: context,
            transformers: [
                (0, dev_server_transformer_1.createDevServerTransformers)(context, this.workspace),
                ...this.transformers,
            ],
        });
    }
    getDevEnvId() {
        const objToHash = {
            webpack: this.webpackModulePath,
            webpackDevServer: this.webpackDevServerModulePath,
            transformers: this.transformers,
        };
        const devEnvId = (0, object_hash_1.default)(objToHash);
        return devEnvId;
    }
    getBundler(context) {
        if (this.bundler)
            return () => this.bundler;
        return webpack_webpack_bundler_1.WebpackBundler.from({
            targets: context.targets,
            transformers: [(0, bundler_transformers_1.createBundlerTransformers)(context), ...this.transformers],
            bundlerContext: context,
            webpackModulePath: this.webpackModulePath,
        });
    }
    getTemplateBundler(context) {
        const bundler = webpack_webpack_bundler_1.WebpackBundler.from({
            targets: context.targets,
            transformers: [
                (0, template_transformers_1.createTemplateBundlerTransformers)(context),
                ...this.transformers,
            ],
            bundlerContext: context,
            webpackModulePath: this.webpackModulePath,
        });
        return bundler;
    }
    getHostDependencies() {
        return this.hostDependencies || ReactPreview.hostDependencies;
    }
    getMounter() {
        return this.mounter;
    }
    getDocsTemplate() {
        return this.docsTemplate;
    }
    getPreviewConfig() {
        return Object.assign({ strategyName: 'component', splitComponentBundle: true }, this.previewConfig);
    }
    static from(options) {
        const name = options.name || 'react-preview';
        return (context) => {
            let workspace;
            try {
                workspace = context.getAspect('teambit.workspace/workspace');
            }
            catch (_a) {
                // ignore - this happen usually when loading the env from a scope when you want to use a workspace template
            }
            return new ReactPreview(name, options.mounter, options.docsTemplate, options.previewConfig, options.transformers, workspace, options.hostDependencies, options.devServer, options.bundler, options.webpackModulePath, options.webpackDevServerModulePath);
        };
    }
}
exports.ReactPreview = ReactPreview;
/**
 * Default host dependencies for the react preview.
 */
ReactPreview.hostDependencies = [
    '@teambit/mdx.ui.mdx-scope-context',
    '@mdx-js/react',
    'react',
    'react-dom',
];
