// @bitdev/react.generators.react-templates@1.1.12/dist/react-env-template/react-env-template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactEnvTemplate = void 0;
const index_file_1 = require("./files/index-file");
const react_env_bit_env_1 = require("./files/react-env-bit-env");
const env_jsonc_1 = require("./files/env-jsonc");
const eslintrc_1 = require("./files/config/eslintrc");
const tsconfig_json_1 = require("./files/config/tsconfig.json");
const jest_config_1 = require("./files/config/jest.config");
const prettier_config_1 = require("./files/config/prettier.config");
const webpack_transformer_1 = require("./files/config/webpack-transformer");
const mounter_1 = require("./files/preview/mounter");
const host_dependencies_1 = require("./files/preview/host-dependencies");
const asset_1 = require("./files/types/asset");
const style_1 = require("./files/types/style");
const docs_1 = require("./files/docs");
class ReactEnvTemplate {
    constructor(name = 'react-env', description = 'create a custom React development environment', hidden = false, isEnv = true, installMissingDependencies = true) {
        this.name = name;
        this.description = description;
        this.hidden = hidden;
        this.isEnv = isEnv;
        this.installMissingDependencies = installMissingDependencies;
    }
    generateFiles(context) {
        return [
            (0, index_file_1.indexFile)(context),
            (0, mounter_1.mounterFile)(),
            (0, host_dependencies_1.hostDependenciesFile)(),
            (0, react_env_bit_env_1.ReactEnv)(context),
            (0, env_jsonc_1.envJson)(),
            (0, eslintrc_1.eslintConfigFile)(),
            (0, docs_1.envDocsFile)(context),
            (0, tsconfig_json_1.tsConfigFile)(),
            (0, jest_config_1.jestConfigFile)(),
            (0, prettier_config_1.prettierConfigFile)(),
            (0, webpack_transformer_1.webpackTransformerFile)(),
            (0, asset_1.assetsFile)(),
            (0, style_1.styleFile)(),
        ];
    }
    config() {
        return {
            'teambit.envs/envs': {
                env: 'teambit.envs/env',
            },
        };
    }
    static from(options = {}) {
        return () => new ReactEnvTemplate(options.name, options.description, options.hidden);
    }
}
exports.ReactEnvTemplate = ReactEnvTemplate;
