// @bitdev/react.generators.react-templates@1.1.12/dist/theme-template/files/theme-file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.themeFile = void 0;
function themeFile(context) {
    return {
        relativePath: `${context.name}.tsx`,
        content: `import { ReactNode } from 'react';
import classNames from 'classnames';
import { ${context.namePascalCase}Provider } from './${context.name}-provider.js';
import styles from './${context.name}.module.scss';

export type ${context.namePascalCase}Props = {
  /**
   * a root ReactNode for the component tree 
   * applied with the theme.
   */
  children?: ReactNode;
  /**
   * inject a class name to override to the theme.
   * this allows people to affect your theme. remove to avoid.
   */
  className?: string;
};

/**
 * a theme for the Acme organization.
 * it provides tokens, fonts and general styling for all components.
 */
export function ${context.namePascalCase}({ children, className, ...rest }: ${context.namePascalCase}Props) {
  return (
    <${context.namePascalCase}Provider.ThemeProvider
      {...rest}
      className={classNames(styles.${context.nameCamelCase}, className)}
    >
      {children}
    </${context.namePascalCase}Provider.ThemeProvider>
  );
}
`
    };
}
exports.themeFile = themeFile;
