// @teambit/webpack.webpack-dev-server@1.0.7/webpack-dev-server.ts
import type { DevServer } from '@teambit/bundler';
import type { Server } from 'http';
import type { Logger } from '@teambit/logger';
import type { Workspace } from '@teambit/workspace';
import type { PubsubMain } from '@teambit/pubsub';
import { join } from 'path';
import findRoot from 'find-root';
import { webpack as webpackCompiler } from 'webpack';
import WDS from 'webpack-dev-server';
import { inspect } from 'util';
import { EnvContext } from '@teambit/envs';
import {
  WebpackConfigDevServerTransformContext,
  WebpackConfigMutator,
} from '@teambit/webpack';
import {
  generateTransformers,
  runTransformers,
} from '@teambit/webpack.webpack-bundler';
import objectHash from 'object-hash';
import { WebpackDevServerOptions } from './webpack-dev-server-options';
import { configFactory } from './webpack.dev.config';
import { WebpackConfigWithDevServer } from './webpack-config-with-dev-server';

export type DevServerAppContext = {
  logger: Logger;
  workspacePath: Workspace['path'];
  pubsub: PubsubMain;
};

type WebpackDevServerAspectsContext = {
  logger: Logger;
  workspacePath: string;
  pubsub: PubsubMain;
};

export class WebpackDevServer implements DevServer {
  private readonly webpack: typeof webpackCompiler;

  private readonly WsDevServer: typeof WDS;

  constructor(
    private config: WebpackConfigWithDevServer,
    private webpackModulePath: string = require.resolve('webpack'),
    private webpackDevServerModulePath: string = require.resolve(
      'webpack-dev-server'
    )
  ) {
    // eslint-disable-next-line import/no-dynamic-require, global-require
    this.webpack = require(this.webpackModulePath);
    // eslint-disable-next-line import/no-dynamic-require, global-require
    this.WsDevServer = require(this.webpackDevServerModulePath);
  }

  private getCompiler(): any {
    return this.webpack(this.config);
  }

  version(): string {
    // Resolve version from the webpack-dev-server package.json
    try {
      const root = findRoot(this.webpackDevServerModulePath);
      const packageJsonPath = join(root, 'package.json');
      // eslint-disable-next-line import/no-dynamic-require, global-require
      const packageJson = require(packageJsonPath);
      return packageJson.version;
    } catch (err) {
      return 'unknown';
    }
  }

  id = 'webpack-dev-server';

  // eslint-disable-next-line react/static-property-placement
  displayName = 'Webpack dev server';

  displayConfig(): string {
    return inspect(this.config, { depth: 10 });
  }

  hash(): string {
    const objToHash = {
      webpack: this.webpackModulePath,
      webpackDevServer: this.webpackDevServerModulePath,
      config: this.config,
    };
    return objectHash(objToHash);
  }

  async listen(port: number): Promise<Server> {
    if (!this.config.devServer) {
      throw new Error('Missing devServer configuration for webpack');
    }
    // Prevent different port between the config port and the listen arg port
    this.config.devServer.port = port;

    // Adding signal listeners to make sure we immediately close the process on sigint / sigterm (otherwise webpack dev server closing will take time)
    this.addSignalListener();

    // Compatibility check for Webpack dev server v3 (e.g. for Angular v8)
    if (
      typeof (this.WsDevServer as any).addDevServerEntrypoints !== 'undefined'
    ) {
      // @ts-ignore in the capsules it throws an error about compatibilities issues between webpack.compiler and webpackDevServer/webpack/compiler
      const webpackDs = new (this.WsDevServer as any)(
        this.getCompiler(),
        this.config.devServer
      );
      return webpackDs.listen(port);
    }

    // @ts-ignore in the capsules it throws an error about compatibilities issues between webpack.compiler and webpackDevServer/webpack/compiler
    const webpackDs: WDS = new this.WsDevServer(
      this.config.devServer,
      this.getCompiler()
    );
    await webpackDs.start();

    return webpackDs.server as Server;
  }

  private addSignalListener() {
    process.on('SIGTERM', () => {
      process.exit();
    });

    process.on('SIGINT', () => {
      process.exit();
    });
  }

  static from(options: WebpackDevServerOptions) {
    return (context: EnvContext) => {
      const name = options.name || 'webpack-dev-server';
      const logger = context.createLogger(name);
      const workspacePath = context.getAspect<any>(
        'teambit.workspace/workspace'
      )?.path;
      const pubsub = context.getAspect<any>('teambit.harmony/pubsub');
      const aspectContext: WebpackDevServerAspectsContext = {
        logger,
        workspacePath,
        pubsub,
      };
      return WebpackDevServer.create(options, aspectContext);
    };
  }

  static create(
    options: WebpackDevServerOptions,
    aspectContext: WebpackDevServerAspectsContext
  ) {
    const { logger, workspacePath, pubsub } = aspectContext;
    const { devServerContext } = options;

    const config = configFactory(
      options.devServerContext.id,
      workspacePath,
      devServerContext.entry,
      devServerContext.rootPath,
      devServerContext.publicPath,
      pubsub
    );

    const configMutator = new WebpackConfigMutator(config);
    const transformerContext: WebpackConfigDevServerTransformContext =
      Object.assign(devServerContext, { mode: 'dev' as const });
    const internalTransformers = generateTransformers(
      logger,
      undefined,
      transformerContext
    );

    const afterMutation = runTransformers(
      configMutator.clone(),
      [...internalTransformers, ...(options.transformers || [])],
      transformerContext
    );

    return new WebpackDevServer(
      afterMutation.raw as WebpackConfigWithDevServer,
      options.webpackModulePath,
      options.webpackDevServerModulePath
    );
  }
}
